src/NexusMods.DataModel/NxFileStore.cs

```diff
@@ -127,44 +126,73 @@ await using (var outputStream = outputPath.Create())
-        Span<byte> buffer = stackalloc byte[24];
+        Span<byte> buffer = stackalloc byte[8];
         BinaryPrimitives.WriteUInt64BigEndian(buffer, hash.Value);
-        guid.TryWriteBytes(buffer.SliceFast(8));
         return IId.FromSpan(EntityCategory.ArchivedFiles, buffer);
     }
 
     /// <inheritdoc />
-    public async Task ExtractFiles(IEnumerable<(Hash Src, AbsolutePath Dest)> files, CancellationToken token = default)
+    public async Task ExtractFiles((Hash Hash, AbsolutePath Dest)[] files, CancellationToken token = default)
     {
-        var grouped = files.Distinct()
-            .Select(input => TryGetLocation(input.Src, out var archivePath, out var fileEntry)
-                ? (true, Hash: input.Src, ArchivePath: archivePath, FileEntry: fileEntry, input.Dest)
-                : default)
-            .Where(x => x.Item1)
-            .ToLookup(l => l.ArchivePath, l => (l.Hash, l.FileEntry, l.Dest));
-
-        if (grouped[default].Any())
-            throw new Exception($"Missing archive for {grouped[default].First().Hash.ToHex()}");
+        // Group the files by archive.
+        // In almost all cases, everything will go in one archive, except for cases
+        // of duplicate files between different mods.
+        var groupedFiles = new ConcurrentDictionary<AbsolutePath, List<(Hash Hash, FileEntry FileEntry, AbsolutePath Dest)>>(Environment.ProcessorCount, 1);
+        var createdDirectories = new ConcurrentDictionary<AbsolutePath, byte>();
+    
+#if DEBUG
+        var destPaths = new ConcurrentDictionary<AbsolutePath, byte>(); // Sanity test. Test code had this issue.
+#endif
+
+        var fileExistsCache = new ConcurrentDictionary<AbsolutePath, bool>(Environment.ProcessorCount, files.Length);
+        Parallel.ForEach(files, file =>
+        {
+            if (TryGetLocation(file.Hash, fileExistsCache, out var archivePath, out var fileEntry))
+            {
+                var group = groupedFiles.GetOrAdd(archivePath, _ => new List<(Hash, FileEntry, AbsolutePath)>());
+                lock (group)
+                {
+                    group.Add((file.Hash, fileEntry, file.Dest));
+                }
 
-        var settings = new UnpackerSettings();
+                // Create the directory, this will speed up extraction in Nx
+                // down the road. Usually the difference is negligible, but in
+                // extra special with 100s of directories scenarios, it can
+                // save a second or two.
+                var containingDir = file.Dest.Parent;
+                if (createdDirectories.TryAdd(containingDir, 0))
+                    containingDir.CreateDirectory();
+#if DEBUG
+                if (!destPaths.TryAdd(file.Dest, 0))
+                    throw new Exception($"Duplicate destination path: {file.Dest}. Should not happen.");
+#endif
+            }
+            else
+            {
+                throw new Exception($"Missing archive for {file.Hash.ToHex()}");
+            }
```

Comment:
Throw a more sensible exception than just `Exception`.

Reply:
Note: This was preserved from the original code. Behaviour wasn't changed here.

I changed it though to throw `FileNotFoundException`.